File: Problems/No.50_Pow(x, n)/pow(x,n).cpp
class Solution {
public:
    
    // Pow(x, n) for n>0
    double positive_power(double x, int n) {
        
        if (n == 1) return x;
        else {
            
            if (n%2 == 0) return positive_power(x*x, n/2);
            else return x*positive_power(x, n-1);
            
        }
        
    }
    
    // Pow(x, n) for n<0
    double negative_power(double x, int n) {
        
        if (n == -1) return x;
        else {
            
            if (n%2 == 0) return negative_power(x*x, n/2);
            else return x*negative_power(x, n+1);
            
        }
        
    }
    
    double myPow(double x, int n) {
        
        if (n == 0) return 1.0;
        else {
            
            if (n > 0) return positive_power(x, n);
            else return negative_power(1/x, n);
            
        }
        
    }
};

File: Problems/No.50_Pow(x,n)/pow(x,n).py
class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        
        # Pow(x, n) for n>0
        def pos_pow(x, n):
            if (n == 1):
                return x
            else:
                if n%2:
                    return x*pos_pow(x, n-1)
                else:
                    return pos_pow(x*x, n//2)
        
        # Pow(x, n) for n<0
        def neg_pow(x, n):
            if (n == -1):
                return x
            else:
                if n%2:
                    return x*neg_pow(x, n+1)
                else:
                    return neg_pow(x*x, n//2)
        
        if not n:
            return 1
        else:
            if (n > 0):
                return pos_pow(x, n)
            else:
                return neg_pow(1/x, n)
